pageserver/src/tenant/timeline.rs

```diff
@@ -2814,30 +2810,36 @@ impl Timeline {
                             .into_iter()
                             .map(|(SearchResult { layer, lsn_floor }, keyspace_accum)| {
                                 (
-                                    ReadableLayerDesc::Persistent {
-                                        desc: (*layer).clone(),
-                                        lsn_range: lsn_floor..cont_lsn,
-                                    },
+                                    ReadableLayer::PersistentLayer(guard.get_from_desc(&layer)),
                                     keyspace_accum.to_keyspace(),
+                                    lsn_floor..cont_lsn,
                                 )
                             })
-                            .for_each(|(layer, keyspace)| fringe.update(layer, keyspace));
+                            .for_each(|(layer, keyspace, lsn_range)| {
+                                fringe.update(layer, keyspace, lsn_range)
+                            });
                     }
                 }
             }
 
-            if let Some((layer_to_read, keyspace_to_read)) = fringe.next_layer() {
+            // It's safe to drop the layer map lock after planning the next round of reads.
+            // The fringe keeps readable handles for the layers which are safe to read even
+            // if layers were compacted or flushed.
+            drop(guard);
```

Comment:
IMO, the fact that a `struct Layer` / `Arc<InMemoryLayer>` is safe to hold outside of the layer map lock needs no special mention. It's implied throughout the code base.

However, what should be explained here is why there's no toctou problem if the layer map changes between iterations. The answer is that we don't rely on the layer map as a whole, nor the layers themselves, but just the logical *contents* beneath `cont_lsn`. (At least I can see myself needing that reminder in the not too distant future :) )

Reply:
> At least I can see myself needing that reminder in the not too distant future

+100

Expanded on the correctness aspect of the comment in https://github.com/neondatabase/neon/pull/7215/commits/b1feb1fa006c3d9526b2bea34fe41509c092082a